gprofiler/profilers/factory.py

```diff
@@ -0,0 +1,45 @@
+from typing import TYPE_CHECKING, Any, List, Tuple, Union
+
+from gprofiler.exceptions import SystemProfilerInitFailure
+from gprofiler.log import get_logger_adapter
+from gprofiler.profilers.profiler_base import NoopProfiler
+
+if TYPE_CHECKING:
+    from gprofiler.main import UserArgs
+    from gprofiler.profilers.profiler_base import ProcessProfilerBase
+
+from gprofiler.profilers.perf import SystemProfiler
+from gprofiler.profilers.registry import get_profilers_registry
+
+logger = get_logger_adapter(__name__)
+COMMON_PROFILER_ARGUMENT_NAMES = ["frequency", "duration"]
+
+
+def get_profilers(
+    user_args: 'UserArgs', **profiler_init_kwargs: Any
+) -> Tuple[Union['SystemProfiler', 'NoopProfiler'], List['ProcessProfilerBase']]:
+    profilers_registry = get_profilers_registry()
+    process_profilers_instances: List['ProcessProfilerBase'] = []
+    system_profiler: Union['SystemProfiler', 'NoopProfiler'] = NoopProfiler()
+    for profiler_name, profiler_config in profilers_registry.items():
+        lower_profiler_name = profiler_name.lower()
+        profiler_mode = user_args.get(f"{lower_profiler_name}_mode")
+        if profiler_mode in ("none", "disabled"):
+            continue
+        try:
+            profiler_kwargs = profiler_init_kwargs.copy()
+            for key, value in user_args.items():
+                if key.startswith(lower_profiler_name) or key in COMMON_PROFILER_ARGUMENT_NAMES:
+                    profiler_kwargs[key] = value
+            profiler_instance = profiler_config.profiler_class(**profiler_kwargs)
+        except Exception:
+            if profiler_config.profiler_class == SystemProfiler:
+                raise SystemProfilerInitFailure("Could not create the system profiler")
```

Comment:
Why swallowing the error here and removing all information attached to the exception?

Do you have any reason for `SystemProfilerInitFailure`? If not, I suggest you remove it and just re-raise the exception. If you have, please attach the original exception values in it

Reply:
1. The exception is _not_ swallowed - the traceback is attached to the new exception and will be logged. 
2. I think it cleans the code, doing `except SystemProfilerInitFailure` to add a new specific log (`System profiler initialization has failed`) is way cleaner than catching all of the exceptions and logging that _something_ failed while attempting to get the profilers. I would not be able to have a precise log otherwise.

Reply:
Right. No `from None` so it will be attached.

Anyway, I ran gProfiler such that `perf` will fail, and I get:
```
2021-08-05 11:00:36,946] ERROR: gprofiler: Running perf failed; consider running gProfiler with '--perf-mode disabled' to avoid using perf
Traceback (most recent call last):
  File "/app/gprofiler/main.py", line 240, in _snapshot
    system_result = system_future.result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/app/gprofiler/profilers/perf.py", line 196, in snapshot
    self._perf_fp.wait_and_script() if self._perf_fp is not None else None,
  File "/app/gprofiler/profilers/perf.py", line 87, in wait_and_script
    perf_data = wait_for_file_by_prefix(f"{self._output_path}.", self._dump_timeout_s, self._stop_event)
  File "/app/gprofiler/utils.py", line 132, in wait_for_file_by_prefix
    wait_event(timeout, stop_event, lambda: len(glob.glob(glob_pattern)) > 0)
  File "/app/gprofiler/utils.py", line 119, in wait_event
    raise TimeoutError()
TimeoutError
[2021-08-05 11:00:36,953] INFO: gprofiler: Stopping ...
[2021-08-05 11:00:36,954] INFO: gprofiler.profilers.perf: Stopped perf (fp mode)
[2021-08-05 11:00:36,954] ERROR: gprofiler: Unexpected error occurred
Traceback (most recent call last):
  File "/app/gprofiler/main.py", line 667, in main
    gprofiler.run_single()
  File "/app/gprofiler/main.py", line 311, in run_single
    self._snapshot()
  File "/app/gprofiler/main.py", line 240, in _snapshot
    system_result = system_future.result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/app/gprofiler/profilers/perf.py", line 196, in snapshot
    self._perf_fp.wait_and_script() if self._perf_fp is not None else None,
  File "/app/gprofiler/profilers/perf.py", line 87, in wait_and_script
    perf_data = wait_for_file_by_prefix(f"{self._output_path}.", self._dump_timeout_s, self._stop_event)
  File "/app/gprofiler/utils.py", line 132, in wait_for_file_by_prefix
    wait_event(timeout, stop_event, lambda: len(glob.glob(glob_pattern)) > 0)
  File "/app/gprofiler/utils.py", line 119, in wait_event
    raise TimeoutError()
TimeoutError
```

which makes sense, since `perf` is only executed on` star` - there's no real reason for `SystemProfiler.__init__` to fail, it doesn't call anything, You can retain the exception tho..